File: server/src/github/github-token-store.ts
import { findGitHubConnectionByLogin, upsertGitHubConnection } from "../db.js";
import { getAdminLogin } from "../config.js";
import { decryptToken, encryptToken } from "./token-crypto.js";

function sessionSecret(): string {
  const secret = process.env.SESSION_SECRET?.trim();
  if (!secret) {
    throw new Error("SESSION_SECRET is not configured");
  }
  return secret;
}

export async function saveGitHubToken(
  login: string,
  githubId: number,
  accessToken: string,
  scopes: string | null | undefined
): Promise<void> {
  const encrypted = encryptToken(accessToken, sessionSecret());
  await upsertGitHubConnection({
    githubId,
    login,
    accessTokenEncrypted: encrypted,
    scopes: scopes ?? "read:user repo",
    updatedAt: Date.now(),
  });
}

export async function loadGitHubToken(login: string): Promise<string | null> {
  const connection = await findGitHubConnectionByLogin(login);
  if (!connection) return null;
  try {
    return decryptToken(connection.access_token_encrypted, sessionSecret());
  } catch {
    return null;
  }
}

export async function resolveStatsToken(): Promise<string | null> {
  const pat = process.env.GITHUB_STATS_TOKEN?.trim();
  if (pat) return pat;
  return loadGitHubToken(getAdminLogin());
}

File: server/Sources/App/Support/GitHubTokenStore.swift
import Vapor

enum GitHubTokenStore {
    static func sessionSecret() throws -> String {
        guard let secret = Environment.get("SESSION_SECRET"), !secret.isEmpty else {
            throw ServiceError.badRequest("SESSION_SECRET is not configured")
        }
        return secret
    }

    static func saveToken(
        for login: String,
        githubId: Int,
        accessToken: String,
        scopes: String?,
        repository: GitHubConnectionRepository
    ) async throws {
        let encrypted = try TokenCrypto.encrypt(accessToken, secret: try sessionSecret())
        _ = try await repository.upsert(
            githubId: githubId,
            login: login,
            accessTokenEncrypted: encrypted,
            scopes: scopes ?? "read:user repo",
            updatedAt: nowMillis()
        )
    }

    static func loadToken(
        for login: String,
        repository: GitHubConnectionRepository
    ) async throws -> String? {
        guard let connection = try await repository.findByLogin(login) else {
            return nil
        }
        return try TokenCrypto.decrypt(connection.accessTokenEncrypted, secret: try sessionSecret())
    }

    static func resolveStatsToken(
        connectionRepository: GitHubConnectionRepository
    ) async throws -> String? {
        if let pat = Environment.get("GITHUB_STATS_TOKEN"), !pat.isEmpty {
            return pat
        }

        let adminLogin = AppConfig.adminLogin()
        return try await loadToken(for: adminLogin, repository: connectionRepository)
    }
}